lebedigital/openbis/expstep.py

```diff
@@ -0,0 +1,1464 @@
+                    collection_dict = {}
+                    for collection in ExpStep.get_collection_names(o, space, project):
+                        sample_list = []
+                        for sample in ExpStep.get_sample_names(o, space, project, collection):
+                            sample_list.append(sample)
+                        collection_dict[collection] = sample_list
+                    project_dict[project] = collection_dict
+                space_dict['DATASTORE'][space] = project_dict
+
+            return space_dict
+
+        elif level == 'space':
+            space = kwargs.pop('space')
+
+            # project_dict = {space: {}}
+            project_dict = {space: {}}
+            for project in ExpStep.get_project_names(o, space):
+                collection_dict = {}
+                for collection in ExpStep.get_collection_names(o, space, project):
+                    sample_list = []
+                    for sample in ExpStep.get_sample_names(o, space, project, collection):
+                        sample_list.append(sample)
+                    collection_dict[collection] = sample_list
+
+                project_dict[space][project] = collection_dict
+
+            return project_dict
+
+        elif level == 'project':
+            space = kwargs.pop('space')
+            project = kwargs.pop('project')
+
+            collection_dict = {project: {}}
+            for collection in ExpStep.get_collection_names(o, space, project):
+                sample_list = []
+                for sample in ExpStep.get_sample_names(o, space, project, collection):
+                    sample_list.append(sample)
+                collection_dict[project][collection] = sample_list
+
+            return collection_dict
+
+        elif level == 'collection':
+            space = kwargs.pop('space')
+            project = kwargs.pop('project')
+            collection = kwargs.pop('collection')
+
+            sample_list = []
+            for sample in ExpStep.get_sample_names(o, space, project, collection):
+                sample_list.append(sample)
+            sample_dict = {collection: sample_list}
+
+            return sample_dict
+
+        else:
+            raise ValueError('No correct level specified')
+
+    @staticmethod
+    def get_props_list(o: Openbis, identifier: str) -> list:
+        return [prop.upper() for prop in list(o.get_sample(identifier).p().keys())]
+
+    def sync_name(self, get_from):
+        """ Synchronizes the name of the attribute with the name in metadata.
+            Have to be the same in order to search for the sample accurately.
+            get_from specifies if the name should be synced from name or metadata:
+            get_from == 'name':  self.name -> self.metadata['$name']
+            get_from == 'metadata':  self.metadata['$name'] -> self.name
+        """
+        if get_from == 'name':
+            self.metadata['$name'] = self.name
```

Comment:
Not sure if I understand that. What is the self.name as opposed to metadata['name']?

Reply:
I am still thinking about the best way to go about that myself. self.name is a name we give the sample but it does not get uploaded into openBIS. metadata["$name"] should be the same thing but can be parsed into a pybis call better. You can search for a sample with self.name better though. If i find a clean solution to that problem i will update the code.